bin/l10n-sync.js

```diff
@@ -0,0 +1,307 @@
+    ).trim();
+
+    console.log(`Found '${upstream}' as upstream remote.`);
+    return upstream;
+  } catch (error) {
+    throw new Error(stripIndent`
+      Couldn't find the upstream remote. Is it well configured?
+      We're looking for either devtools-html/perf.html or firefox-devtools/profiler.
+    `);
+  }
+}
+
+/**
+ * Compares the `compareBranch` with `baseBranch` and checks the changed files.
+ * Fails if the `compareBranch` has changes from the files that doesn't match
+ * the `matchRegexp`.
+ *
+ * @param {string} upstream Name of the upstream remote.
+ * @param {string} compareBranch Branch that we want to compare.
+ * @param {string} baseBranch Branch that we want to take as base for comparison.
+ * @param {RegExp} matchRegexp Regexp that will be used to match the files.
+ * @throws Will throw an error if `compareBranch` has changes from the files
+ * that doesn't match the `matchRegexp`.
+ */
+async function checkAllowedPaths(
+  upstream,
+  compareBranch,
+  baseBranch,
+  matchRegexp
+) {
+  console.log(
+    `>>> Checking if ${compareBranch} branch has changes from the files that are not allowed.`
+  );
+  // git log --no-merges compareBranch..baseBranch --pretty="format:" --name-only
+  let changedFiles = await logAndExec(
+    'git',
+    'log',
+    '--no-merges',
+    `${upstream}/${compareBranch}..${upstream}/${baseBranch}`,
+    '--pretty=format:',
+    '--name-only'
+  );
+  changedFiles = changedFiles.split('\n');
+
+  for (const file of changedFiles) {
+    if (file.length > 0 && !file.match(matchRegexp)) {
+      throw new Error(oneLine`
+        ${compareBranch} branch includes changes from the files that are not
+        allowed: ${file}
+      `);
+    }
+  }
+}
+
+/**
+ * This is a simple helper function that returns the friendly English versions
+ * of how many times that occurs.
+ *
+ * It's a pretty simple hack and would be good to have a more sophisticated
+ * (localized?) API function. But it's not really worth for a deployment only
+ * script.
+ *
+ * @param {number} count
+ * @returns {string}
+ */
+function fewTimes(count) {
+  switch (count) {
+    case 1:
+      return 'once';
+    case 2:
+      return 'twice';
+    default:
+      return `${count} times`;
+  }
+}
+
+/**
+ * Tries to sync the l10n branch and retries for 3 times if it fails to sync.
+ *
+ * @param {string} upstream
+ * @returns {Promise<void>}
+ * @throws Will throw an error if it fails to sync for more than 3 times.
+ */
+async function tryToSync(upstream) {
+  console.log('>>> Syncing the l10n branch with main.');
+  // RegExp for matching only the vendored locales.
+  const vendoredLocalesPath = /^locales[\\/](?!en-US[\\/]).*$/g;
+  // RegExp for matching anything but the vendored locales.
+  const nonVendoredLocalesPath = /^(locales[\\/]en-US[\\/]|(?!locales[\\/])).*$/g;
+  const totalTryCount = 3;
+  let error = null;
+  let tryCount = 0;
+
+  // Try to sync and retry for `totalTryCount` times if it fails.
+  do {
```

Comment:
I'm always happy when I use `do .. while` because this is so rare 😁

Reply:
(I'm thinking we could use a `while(true)` with `break` and `continue` in the right places, but `do .. while` is so satisfying)